go/plugins/ollama/ollama.go

```diff
@@ -0,0 +1,340 @@
+	return ai.LookupModel(provider, name)
+}
+
+// ModelDefinition represents a model with its name and type.
+type ModelDefinition struct {
+	Name string
+	Type string
+}
+
+// Config provides configuration options for the Init function.
+type Config struct {
+	// Server Address of oLLama.
+	ServerAddress string
+	// Generative models to provide.
+	Models []ModelDefinition
+}
+
+// Init registers all the actions in this package with ai.
+func Init(ctx context.Context, cfg Config) error {
+	for _, model := range cfg.Models {
+		defineModel(model, cfg.ServerAddress)
+	}
+	return nil
+}
+
+type generator struct {
+	model         ModelDefinition
+	serverAddress string
+}
+
+type ollamaMessage struct {
+	Role    string // json:"role"
+	Content string // json:"content"
+}
+
+/*
+TODO: Support optional, advanced parameters:
+format: the format to return a response in. Currently the only accepted value is json
+options: additional model parameters listed in the documentation for the Modelfile such as temperature
+system: system message to (overrides what is defined in the Modelfile)
+template: the prompt template to use (overrides what is defined in the Modelfile)
+context: the context parameter returned from a previous request to /generate, this can be used to keep a short conversational memory
+stream: if false the response will be returned as a single response object, rather than a stream of objects
+raw: if true no formatting will be applied to the prompt. You may choose to use the raw parameter if you are specifying a full templated prompt in your request to the API
+keep_alive: controls how long the model will stay loaded into memory following the request (default: 5m)
+*/
+type ollamaRequest struct {
+	Messages []*ollamaMessage `json:"messages"`
+	Model    string           `json:"model"`
+	Stream   bool             `json:"stream"`
+}
+
+type ollamaGenerateRequest struct {
+	System string `json:"system,omitempty"` // Optional System field
+	Model  string `json:"model"`
+	Prompt string `json:"prompt"`
+	Stream bool   `json:"stream"`
+}
+
+// TODO: Add optional parameters (images, format, options, etc.) based on your use case
+type ollamaResponse struct {
+	Model     string `json:"model"`
+	CreatedAt string `json:"created_at"`
+	Message   struct {
+		Role    string `json:"role"`
+		Content string `json:"content"`
+	} `json:"message"`
+}
+
+type ollamaGenerateResponse struct {
+	Model     string `json:"model"`
+	CreatedAt string `json:"created_at"`
+	Response  string `json:"response"`
+}
+
+// Generate makes a request to the Ollama API and processes the response.
+func (g *generator) generate(ctx context.Context, input *ai.GenerateRequest, cb func(context.Context, *ai.GenerateResponseChunk) error) (*ai.GenerateResponse, error) {
+
+	stream := cb != nil
+	var payload any
+	isChatModel := g.model.Type == "chat"
+	if !isChatModel {
+		payload = ollamaGenerateRequest{
+			Model:  g.model.Name,
+			Prompt: concatMessages(input, ai.Role("user")),
```

Comment:
This feels strange that we're concatenating all the same-role text. It loses the interleaving of the conversation so far.
Does the JS ollama client do this?

Reply:
True... there are some issues with the ollama JS client as well (doesn't have a system prompt), but you're right that it makes more sense to interleave all the messages. (at least the non-system messages)